### PR TITLE
Added emailing of detected anomalies

### PR DESCRIPTION
**Pull request by thibautcornolti**
Added task `anomalies-detection` checking anomalies of the past week and emailing them to team@trackit.io.
An anomaly can't be emailed twice.

### CODE REVIEW

File: aws/aws.go [hunk 1]
```diff
@@ -26,6 +26,7 @@ import (
 	"github.com/trackit/trackit-server/awsSession"
 	"github.com/trackit/trackit-server/models"
 	"github.com/trackit/trackit-server/users"
+	"regexp"
```
Comment (Lolincolc) [inline aws/aws.go]: I think that "regexp" should be placed at the top of import block.

File: db/migration/0010_add_last_emailed_anomalies.sql [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+--
+--   Unless required by applicable law or agreed to in writing, software
+--   distributed under the License is distributed on an "AS IS" BASIS,
+--   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+--   See the License for the specific language governing permissions and
+--   limitations under the License.
+
+CREATE TABLE emailed_anomaly (
+	id                     INTEGER      NOT NULL AUTO_INCREMENT,
+	aws_account_id 				 INTEGER      NOT NULL,
```
Comment (simonmeyerrr) [inline db/migration/0010_add_last_emailed_anomalies.sql]: bad alignment, use spaces and not tab

File: aws/aws.go (hunk 1 above)
Comment (simonmeyerrr) [inline aws/aws.go]: this import should not be at this place

File: db/schema.sql [hunk 3]
```diff
@@ -253,3 +253,26 @@ ALTER TABLE aws_bill_repository ADD grace_update DATETIME NOT NULL DEFAULT "1970
...
+--
+--   Unless required by applicable law or agreed to in writing, software
+--   distributed under the License is distributed on an "AS IS" BASIS,
+--   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+--   See the License for the specific language governing permissions and
+--   limitations under the License.
+
+CREATE TABLE emailed_anomaly (
+	id                     INTEGER      NOT NULL AUTO_INCREMENT,
+	aws_account_id 				 INTEGER      NOT NULL,
```
Comment (simonmeyerrr) [inline db/schema.sql]: same

File: models/awsaccount.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
+//   Copyright 2017 MSolution.IO
```
Comment (simonmeyerrr) [inline models/awsaccount.go]: we are in 2018

File: models/emailedanomaly.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
+//   Copyright 2017 MSolution.IO
```
Comment (simonmeyerrr) [inline models/emailedanomaly.go]: we still are in 2018

File: server/taskAnomaliesDetection.go [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+package main
+
+import (
+	"context"
+	"database/sql"
+	"github.com/trackit/jsonlog"
+	"github.com/trackit/trackit-server/db"
+	"github.com/trackit/trackit-server/models"
+	"github.com/trackit/trackit-server/costs/anomalies"
+	"time"
```
Comment (simonmeyerrr) [inline server/taskAnomaliesDetection.go]: import is not well ordered

File: server/taskAnomaliesDetection.go [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+	"github.com/trackit/trackit-server/aws"
+	"github.com/trackit/trackit-server/users"
+	"github.com/trackit/trackit-server/mail"
+	"fmt"
```
Comment (simonmeyerrr) [inline server/taskAnomaliesDetection.go]: same

File: models/emailedanomaly.go (hunk 5 above)
Comment (simonmeyerrr) [inline models/emailedanomaly.go]: we still are in 2018

File: models/awsaccount.go (hunk 4 above)
Comment (simonmeyerrr) [inline models/awsaccount.go]: we are in 2018

File: db/migration/0010_add_last_emailed_anomalies.sql (hunk 2 above)
Comment (simonmeyerrr) [inline db/migration/0010_add_last_emailed_anomalies.sql]: bad alignment, use spaces and not tab

File: db/schema.sql (hunk 3 above)
Comment (simonmeyerrr) [inline db/schema.sql]: same

File: aws/aws.go (hunk 1 above)
Comment (simonmeyerrr) [inline aws/aws.go]: this import should not be at this place

File: server/taskAnomaliesDetection.go [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+		Which account? %s
+		Which product? %s
+		How much? %f
+		How important? %.1f
+		When? %s</p>
+See more on our website http://re.trackit.io/
+
+this email is originally intended for %s.
+`, awsAccount.Pretty, product, an.Cost, an.Cost / an.UpperBand * 100, date.String(), ea.Recipient)
+	var recipient = "team@trackit.io" // replace by ea.Recipient.
```
Comment (simonmeyerrr) [inline server/taskAnomaliesDetection.go]: did you forget to change it or is it for prod deployement?

File: server/taskAnomaliesDetection.go [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+// sendAnomalyEmail actually sends the mail and log everything.
+func sendAnomalyEmail(user users.User, awsAccount aws.AwsAccount, product string, an anomalies.CostAnomaly, date time.Time, ctx context.Context) error {
+	logger := jsonlog.LoggerFromContextOrDefault(ctx)
+	ea := models.EmailedAnomaly{
+		Date:         date,
+		Recipient:    user.Email,
+		AwsAccountID: awsAccount.Id,
+		Product:      product,
+	}
+	var subject = fmt.Sprintf("TrackIt detected an abnormal peak in your %s cost on your account %s!", product, awsAccount.Pretty)
```
Comment (simonmeyerrr) [inline server/taskAnomaliesDetection.go]: why use everywhere in this function `var ... = ...` and not `... := ...`

**Review by simonmeyerrr — CHANGES_REQUESTED** [review]
there are several changes to be made

File: server/taskAnomaliesDetection.go [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+	identity, err := awsAccount.GetAwsAccountIdentity()
+	if err != nil {
+		logger.Error("Error when getting identity from AwsAccount", map[string]interface{}{
+			"awsAccount": awsAccount,
+			"error":      err,
+		})
+		return err
+	}
+	params := anomalies.AnomalyEsQueryParams{
+		DateBegin:   time.Now().Add(-1 * 24 * 7 * time.Hour),
```
Comment (zanchi-r) [inline server/taskAnomaliesDetection.go:124]: Keep in mind that the tasks are not started at a fixed time, we will miss data if we request only the last week. We should start on the time of the last anomaly detection

**Review by zanchi-r — DISMISSED** [review]

File: models/emailedanomaly.go (hunk 5 above)
Comment (thibautcornolti) [reply models/emailedanomaly.go]: Are we? :thinking:

File: server/taskAnomaliesDetection.go (hunk 10 above)
Comment (thibautcornolti) [reply server/taskAnomaliesDetection.go:124]: The task has to be launched at least every week so we won't miss anything.
We could miss an anomaly if we launch it every 2 weeks for example.
The goal is to launch it every 12h so the anomalies are quickly detected and emailed.

Comment (zanchi-r) [reply server/taskAnomaliesDetection.go:124]: ok!

**Review by simonmeyerrr — APPROVED** [review]

**Pull request merged**